src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs

```diff
@@ -30,5 +33,51 @@ public static ImmutableArray<Document> GetVisibleDocuments(this IDocumentTrackin
                       .WhereNotNull()
                       .Distinct()
                       .ToImmutableArray();
+
+        /// <summary>
+        /// Waits the specified amount of time while the specified <paramref name="document"/> is not visible.  If any
+        /// document visibility changes happen, the delay will cancel.
+        /// </summary>
+        public static Task DelayWhileNonVisibleAsync(this IDocumentTrackingService service, Document document, TimeSpan timeSpan, CancellationToken cancellationToken)
+            => DelayWhileNonVisibleAsync(service, ImmutableArray.Create(document), timeSpan, cancellationToken);
+
+        public static async Task DelayWhileNonVisibleAsync(this IDocumentTrackingService service, ImmutableArray<Document> documents, TimeSpan timeSpan, CancellationToken cancellationToken)
+        {
+            // Only add a delay if we have access to a service that will tell us when the documents become visible or not.
+            if (!service.SupportsDocumentTracking)
+                return;
+
+            // ensure we listen for visibility changes before checking.  That way we don't have a race where we check
+            // something see it is not visible, but then do not hear about its visibility change because we've hooked up
+            // our event after that happens.
+            var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
+            service.ActiveDocumentChanged += OnActiveDocumentChanged;
+
+            try
+            {
+                // if any of the docs are visible, then no delay.
+                var visibleDocuments = service.GetVisibleDocuments();
+                var anyVisible = documents.Select(d => d.Id).Intersect(visibleDocuments).Any();
+                if (anyVisible)
+                    return;
+
+                // Listen to when the active document changed so that we startup work on a document once it becomes visible.
+
+                var delayTask = Task.Delay(timeSpan, cancellationToken);
```

Comment:
❗ This needs to use `IExpeditableDelaySource.Delay`

Reply:
https://github.com/dotnet/roslyn/pull/60608